glib/src/value.rs

```diff
@@ -80,44 +76,216 @@
 //! ```
 
 use libc::{c_char, c_void};
-use std::borrow::Borrow;
 use std::error;
 use std::ffi::CStr;
 use std::fmt;
-use std::marker::PhantomData;
 use std::mem;
 use std::ops::Deref;
 use std::ptr;
 
 use crate::gstring::GString;
 use crate::translate::*;
 use crate::types::{StaticType, Type};
+use crate::{ParamFlags, ParamSpec};
+
+/// A type that can be stored in `Value`s.
+pub trait ValueType: ToValue + for<'r> FromValue<'r> /* TODO: TryFrom<Value> + Into<Value> */ {
+    /// Type to get the `Type` from.
+    ///
+    /// This exists only for handling nullable types.
+    // FIXME: Should default to Self once associated type defaults are stabilized
+    // https://github.com/rust-lang/rust/issues/29661
+    type Type: StaticType;
+}
+
+/// A type that can be stored in `Value`s and is nullable.
+///
+/// These are types were storing an `Option` is valid. Examples are `String` and all object types.
+pub trait ValueTypeNullable:
+    ValueType + ToValueNullable + for<'r> FromValue<'r, Error = WrongValueTypeOrNoneError>
+{
+}
+
+impl<'r, T: ValueTypeNullable> ValueType for Option<T>
+where
+    Self: for<'s> FromValue<'s, Error = WrongValueTypeError>,
+{
+    type Type = T::Type;
+}
+
+/// A type for which a `ParamSpec` exists and that can be used for object properties.
+pub trait HasParamSpec: ValueType {
```

Comment:
Nice! (also the following ones oc).

Minor nit though :) Since this is more about a type that allows defining `ParamSpec`s with specificities, `Has` might suggest that a particular `ParamSpec` is associated with the type. Maybe we could use `CanParamSpec` instead (and `CanParamSpecWithDefault`, ...)

Reply:
`Create` or `Construct` maybe? It's not that this value "can" param spec, but you can "create"/"construct" one. Seems like a better verb or what do you think?

Reply:
Oc! I only tried something short so far. I'll come back here if I find something else.

Reply:
How about `UseableAsParam`,  `UseableAsParamWithDefault` & `UseableAsParamWithMinMax`? I think this captures the meaning better than the `ParamSpec` orientation. We could also use `CanUseAsParam*`, shorted :)

Reply:
`UseableAsParam` it is